src/main/java/de/tum/in/www1/artemis/service/exam/ExamLiveEventsService.java

```diff
@@ -0,0 +1,117 @@
+        this.studentExamRepository = studentExamRepository;
+    }
+
+    /**
+     * Sends an announcement to all students in the exam.
+     *
+     * @param exam    the exam to send the announcement to.
+     * @param message The message to send.
+     * @param sentBy  The user who sent the message.
+     * @return The created event.
+     */
+    public ExamWideAnnouncementEvent createExamAnnouncementEvent(Exam exam, String message, User sentBy) {
+        var event = new ExamWideAnnouncementEvent();
+
+        // Common fields
+        event.setExamId(exam.getId());
+        event.setCreatedBy(sentBy.getName());
+
+        // Specific fields
+        event.setTextContent(message);
+
+        return this.storeAndDistributeLiveExamEvent(event);
+    }
+
+    /**
+     * Send a working time update to the specified student.
+     *
+     * @param studentExam    The student exam the time was updated for
+     * @param newWorkingTime The new working time
+     * @param oldWorkingTime The old working time
+     * @param courseWide     set to true if this event is caused by a course wide update that affects all students; false otherwise
+     * @param sentBy         The user who performed the update
+     */
+    public void createWorkingTimeUpdateEvent(StudentExam studentExam, int newWorkingTime, int oldWorkingTime, boolean courseWide, User sentBy) {
+        var event = new WorkingTimeUpdateEvent();
+
+        // Common fields
+        event.setExamId(studentExam.getExam().getId());
+        event.setStudentExamId(studentExam.getId());
+        event.setCreatedBy(sentBy.getName());
+
+        // Specific fields
+        event.setNewWorkingTime(newWorkingTime);
+        event.setOldWorkingTime(oldWorkingTime);
+        event.setCourseWide(courseWide);
+
+        this.storeAndDistributeLiveExamEvent(event);
+    }
+
+    /**
+     * Stores the event in the database and sends it either to all student exams of the exam or to a specific student exam,
+     * depending on whether the event is for a specific student exam or for the entire exam.
+     *
+     * @param event The event to store and distribute.
+     * @return The stored event.
+     */
+    private <T extends ExamLiveEvent> T storeAndDistributeLiveExamEvent(T event) {
+        var storedEvent = examLiveEventRepository.save(event);
+
+        // If the event is for a specific student exam, only send it to that student exam.
+        if (event.getStudentExamId() != null) {
+            websocketMessagingService.sendMessage(STUDENT_EXAM_EVENT.formatted(storedEvent.getStudentExamId()), storedEvent.asDTO());
+            return storedEvent;
+        }
+
+        // Otherwise, send it to all student exams of the exam.
+        studentExamRepository.findAllIdsWithoutTestRunsByExamId(storedEvent.getExamId())
+                .forEach(studentExamId -> websocketMessagingService.sendMessage(STUDENT_EXAM_EVENT.formatted(studentExamId), storedEvent.asDTO()));
```

Comment:
could we send this one one topic? this might improve performance

Reply:
Yeah, I guess - this would also increase the number of subscriptions, though